### ISSUE TITLE
[Feature]: Limit thinking tokens

### ISSUE THREAD

**Issue by gohar94** [vllm-project/vllm#15418]
### 🚀 The feature, motivation and pitch

Is there plan to support an option (e.g., `max_thinking_tokens`) to limit the thinking tokens for reasoning models? Perhaps both a minimum and a maximum limit would be useful. This is an example of one approach to achieving this: https://simonwillison.net/2025/Jan/22/r1py/

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by csvance** [comment]
If I understand these models correctly, it should be possible to already to do this with two separate generation prompts / LLM calls, albeit with the limitation that the limit is characters not tokens. You will have to turn off vLLM reasoning model support and use your own parser on the application side.

First prompt starts with the assistant message "<reasoning>" and uses regex constrained output like r".{0,1024}</thinking>". Then the second chat template starts with assistant message "<reasoning>{reasoning_content}</reasoning>" for generation.

Of course it would be nice to be able to not have to work against the built in reasoning model support to do this.

**Comment by yelaman-starrez** [comment]
I have come up with logits processor approach:

```
from typing import List, Optional
import torch

class ThinkLogitsProcessor:
    """A logits processor that limit the number of thinking tokens."""
    
    def __init__(self, think_start_token, think_end_token, num_think_tokens: int = 100):
        """
        Initialize the think logits processor.
        
        Args:
            tokenizer: The tokenizer used for the model
            num_think_tokens: Maximum number of tokens allowed in thinking section
        """
        self.num_think_tokens = num_think_tokens
        self.think_start_token = think_start_token
        self.think_end_token = think_end_token
        
    def __call__(
        self,
        input_ids: List[int],
        logits: torch.Tensor,
    ) -> torch.Tensor:
        """
        Process the logits to enforce </think> token when needed.
        
        Args:
            input_ids: List of input token IDs.
            scores: Tensor of logits for the next token.
            
        Returns:
            Processed logits tensor.
        """
        # Check if we're in a thinking section
        if self.think_start_token in input_ids and self.think_end_token not in input_ids:
            # Find the position of the last <think> token
            think_start_pos = len(input_ids) - 1 - input_ids[::-1].index(self.think_start_token)
            
            # Calculate number of tokens since <think>
            tokens_since_think = len(input_ids) - think_start_pos - 1
            
            # If we've reached the maximum thinking length, force </think>
            if tokens_since_think >= self.num_think_tokens:
                # Set all other logits to -inf except for </think>
                logits = torch.full_like(logits, float('-inf'))
                logits[self.think_end_token] = 1.0
                
        return logits 
```

And then use it in sampling params:
```
sampling_params = SamplingParams(
            temperature=temperature,
            max_tokens=max_tokens,
            logits_processors=[ThinkLogitsProcessor(think_start_token, think_end_token, num_think_tokens)]
        )
```

**Comment by Kenwwww** [comment]
> I have come up with logits processor approach:
> 
> ```
> from typing import List, Optional
> import torch
> 
> class ThinkLogitsProcessor:
>     """A logits processor that limit the number of thinking tokens."""
>     
>     def __init__(self, think_start_token, think_end_token, num_think_tokens: int = 100):
>         """
>         Initialize the think logits processor.
>         
>         Args:
>             tokenizer: The tokenizer used for the model
>             num_think_tokens: Maximum number of tokens allowed in thinking section
>         """
>         self.num_think_tokens = num_think_tokens
>         self.think_start_token = think_start_token
>         self.think_end_token = think_end_token
>         
>     def __call__(
>         self,
>         input_ids: List[int],
>         logits: torch.Tensor,
>     ) -> torch.Tensor:
>         """
>         Process the logits to enforce </think> token when needed.
>         
>         Args:
>             input_ids: List of input token IDs.
>             logits: Tensor of logits for the next token.
>             
>         Returns:
>             Processed logits tensor.
>         """
>         # Check if we're in a thinking section
>         if self.think_start_token in input_ids and self.think_end_token not in input_ids:
>             # Find the position of the last <think> token
>             think_start_pos = len(input_ids) - 1 - input_ids[::-1].index(self.think_start_token)
>             
>             # Calculate number of tokens since <think>
>             tokens_since_think = len(input_ids) - think_start_pos - 1
>             
>             # If we've reached the maximum thinking length, force </think>
>             if tokens_since_think >= self.num_think_tokens:
>                 # Set all other logits to -inf except for </think>
>                 logits = torch.full_like(logits, float('-inf'))
>                 logits[self.think_end_token] = 1.0
>                 
>         return logits 
> ```
> 
> And then use it in sampling params:
> 
> ```
> sampling_params = SamplingParams(
>             temperature=temperature,
>             max_tokens=max_tokens,
>             logits_processors=[ThinkLogitsProcessor(think_start_token, think_end_token, num_think_tokens)]
>         )
> ```

how to use it in openai server？

**Comment by csvance** [comment]
Being able to limit thinking tokens would allow my application to use the same LLM for utility tasks as it does for generation. The amount of time a model might spend thinking can vary wildly especially when using constrained output. You might get paragraphs of tokens during thinking only to generate a simple `r'(Yes|No)' `. Not having control over this means I have to tie up more GPU memory for additional models.

**Comment by dhruvmullick** [comment]
I tried a similar logit processor approach, but this leads to slowdowns because vLLM v1 doesn't support logit processors yet.

**Comment by renjie0** [comment]
> I tried a similar logit processor approach, but this leads to slowdowns because vLLM v1 doesn't support logit processors yet.

Why does it cause slowdown? Is it hard to make logit processor run fast in v1?

**Comment by samsja** [comment]
I also think this would be a nice feature to have following this paper https://arxiv.org/abs/2505.05315

**Comment by aarnphm** [comment]
Once we have https://github.com/vllm-project/vllm/pull/16728 merged, this should be easier to be implemented here. We can probably use this to mimic OpenAI's `reasoning_effort` feature 🤔

**Comment by llsj14** [comment]
I’m currently implementing this feature, and I’ve encountered some problems as follows.

First, ~extending the builtin LogitsProcessor has been challenging because its apply method doesn’t take `sampling_metadata` as an argument, so it’s not straightforward to handle input or output tokens within the current design.~ 
I found that I can use a reference to the output token IDs, so I may need to add a reference to the input token IDs in the same way.

Passing reasoning token IDs from the reasoning parser and tokenizer to the LogitsProcessor through the InputBatch also feels misaligned with the current codebase’s philosophy and might require some refactoring.

Second, we could consider token-level limits for DeepSeek-R1 or Qwen reasoning models. However, after reviewing the Graphite reasoning parser, I found that its reasoning tokens are not represented as single tokens. So for now, I’m limiting the thinking tokens only for DeepSeek-R1 and Qwen models.

I hope to share my drafts soon and get your feedback.
